Refactor views and server configuration

Boluwatife-AJB opened the pull request:
- Adjusted the head.ejs file to remove unnecessary closing HTML tag.
- Removed extra line in server.js file.
- Updated app.js file to set the views directory and add routes.
- Added axios as a dependency in package.json.
- Added sign-in and index JavaScript files for client-side functionality.
- Updated README.md to include reset-password functionality.
- Added auth-layout.ejs file for shared layout in authentication pages.
- Created forgot-password.ejs and reset-password.ejs files for password reset functionality.
@opeolluwa

Boluwatife-AJB committed:
Refactor views and server configuration

- Adjusted the head.ejs file to remove unnecessary closing HTML tag.
- Removed extra line in server.js file.
- Updated app.js file to set the views directory and add routes.
- Added axios as a dependency in package.json.
- Added sign-in and index JavaScript files for client-side functionality.
- Updated README.md to include reset-password functionality.
- Added auth-layout.ejs file for shared layout in authentication pages.
- Created forgot-password.ejs and reset-password.ejs files for password reset functionality.

opeolluwa commented on 02_jwt_auth_flow/public/js/sign-in.js:
```diff
@@ -0,0 +1,17 @@
+
+export const signIn = async (email, password) => {
+  try {
+    const response = await axios.post("127.0.0.1:8080/api/auth/signin", {
```
Since it is the same website, you can leave out the base URL, this would resolve to "/api/auth/signin"

Boluwatife-AJB committed:
feat:implemented the sign in functionality to sign users in the client side

opeolluwa commented on 02_jwt_auth_flow/app.js:
```diff
@@ -16,6 +16,9 @@ app.use(express.json());
+// app.use(express.static(`${__dirname}/public`));
+
+
 // Routes
```
this comment may be redundant by just saying "routes", Perhaps you want to provide additional context
```suggestion
/**
* the routes mount point"
* all the view layouts are mounted at / path 
* and the controllers are mounted at /api/auth
* @see /routes for more details 
```

opeolluwa replied:
See https://jsdoc.app/ for best practices on documenting JavaScript code. Finally commenting should be approached with the idea of explaining one's code to someone, I added some examples below

```js 

import fs from "node:fs/promises";
import { build } from "./build.js";
import { EXERCISE_V2_SOURCE_DIRECTORY } from "./index.js";
import { log_error_message } from "./utils.js";

/**
 * @function build_all_exercises - compiles asynchronously for production mode 
 *
 * @params void
 *
 * @returns void
 */

export async function build_all_exercises() {
  const exercises = await fs
    .readdir(EXERCISE_V2_SOURCE_DIRECTORY)
    .catch((error) => {
      log_error_message("Build failed due to " + error.message);
    });
  for (const entry of exercises) {
    /**
     * ignore all "dotfiles" found in the directory 
     * this could be files like .git, .env or .DS_store for Apple laptops 
     */
    if (!entry.startsWith(".")) {
      build(entry);
    }
  }
}
```

opeolluwa commented on 02_jwt_auth_flow/app.js:
```diff
@@ -16,6 +16,9 @@ app.use(express.json());
+
 // Routes
 app.use("/", viewRoutes);
 app.use("/api/auth", authRoutes);
```
There is a modification that can be made to this, and that is API versioning. API should be versioned to help frontend application work perperly with it and track changes. 
API versioning is mostly important to cases of a major changes in the REQUEST or RESPONSE objects; instead of a modifying the existing endpoint a new one is made

Some common versioning strategies are:
`/api/auth/v1/login`
`/api/auth/login?version=1`
`/api/1/auth/login`

The last one isn't very developer or user friendly because numbers like `1`, `2` ... can be confusing without more context.

Can't remeber where I read it, I'll see if I cant the resource across.

opeolluwa commented:
![Screenshot 2024-10-05 at 11 28 34](https://github.com/user-attachments/assets/38d8479c-7a68-4ffb-8e57-a51df757e5ae)

Does the view work correctly?

opeolluwa commented:
![Uploading Screenshot 2024-10-05 at 11.28.43.png…]()

Boluwatife-AJB commented:
@opeolluwa  the active view routes are `/sign-in,` `/sign-up`, `/forgot-password`, `/verify-password`, `/reset-password`.

Boluwatife-AJB committed:
feat: implement complete authentication and password reset flow

- Refactor `signIn` and `signUp` functions to improve error handling and user feedback
- Add new features for `forgotPassword`, `verifyOtp`, and `resetPassword` flows
- Store and retrieve reset password token from localStorage for verification and reset
- Add form handling and delegation for sign-in, sign-up, password recovery, and OTP verification
- Implement password visibility toggle for password and confirm password fields
- Enhance OTP input handling with focus management and backspace navigation

Boluwatife-AJB committed:
feat: add new feature for resend otp functionality
